### ISSUE TITLE
No callback for detecting "Answered" connection state

### ISSUE THREAD

**Issue by davidvoiq** [twilio/twilio-client-quickstart-android#39]
Hi, can you please provide a callback in your mobiles SDK's for being notified when an outgoing connection from the phone has been answered by the person being called?

I've been checking the Twilio's Android SDK logs in verbose mode and it seems the Signaling process detects all those states (initiated, ringing, answered and completed), so I think there is only need for create a way of notify to the java layer those events.

Thanks

**Comment by kbagchiGWC** [comment]
@davidvoiq we don't have a callback in our mobile SDKs to indicate when the remote party answers. However, we do support these events reporting via `statusCallbackEvent`. With `statusCallbackEvent`, you can subscribe to receive webhooks for the different call progress events: `initiated`, `ringing`, `answered`, or `completed` for a given call. For details please refer to the documentation in https://www.twilio.com/docs/api/twiml/client and https://www.twilio.com/docs/api/twiml/number.

**Comment by davidvoiq** [comment]
@kbagchiGWC so what would be the way for starting a timer in our mobile apps as soon the person being called answers the call? 
We don't have any kind of socket communication from/to our server and don't trust in push notifications for sending one push for the device, and also don't wan't to be polling our server each xx milliseconds because the overhead that this can cause in our server.

I think this kind of callback (notifying changes in the connection's state) should be as basic and required in a mobile SDK as well as the [ConnectionListener](https://www.twilio.com/docs/api/client/android/connectionlistener), can you Twilio please create this callback? 

Thank you very much

**Comment by davidvoiq** [comment]
Hi @kbagchiGWC can you please take a look at my previous comment? Thank you very much

**Comment by kbagchiGWC** [comment]
Hi @davidvoiq I agree with you that having those callback as part of `ConnectionListener` will e ideal. But that is not what the SDK supports at the moment. We are working with our backend team to find a long term solution. 

You may find our [blog](https://www.twilio.com/blog/2015/05/introducing-call-progress-events-flexibly-track-and-control-your-outbound-calls.html) that details how to receive these events using webhooks useful.

**Comment by davidvoiq** [comment]
Thank you very much @kbagchiGWC hope the engineering team release this feature soon!
But returning to the original question and summarizing, by now the only way to get noticed of these events from the mobile side is to poll each xx milliseconds our server (wich should keep records of the states of all the calls our system do) from the iOS/Android apps querying for a specific callSid?

**Comment by idelgado** [comment]
@davidvoiq For now I would recommend the following to learn about the answered event.

1a Send the to/from to your server to provide a call record.
1b Or you can wait until the connected event and send the call SID which will be the parent SID provided in the statusCallbackEvent later.
2 Use a websocket, socket or a synchronization service(firebase, etc.) between the client and your server
3 When placing the call subscribe to the statusCallbackEvent by setting the following in the TwiML parameters when you place the call
````
StatusCallback="https://www.myapp.com/events"
StatusCallbackEvent="initiated ringing answered completed"
````
4 When the call is answered your server will receive the answered event along with the to, from, and parent SID of the placed call.
5 With the established websocket, socket, or synchronization service between your server and your client, notify your client of the events.

**Comment by davidvoiq** [comment]
Hi @idelgado, those are in fact the additions we want to avoid to implement in our mobile apps and webserver just to bypass this kind of feature-lack from our third party SDK's.

**Comment by idelgado** [comment]
@davidvoiq Yeah I totally understand. It's definitely a gap we want to provide a simple SDK based solution for. Im still working with the Voice infrastructure team on this.

**Comment by davidvoiq** [comment]
Thanks @idelgado, we'll be eagerly waiting that feature

**Comment by davidvoiq** [comment]
Hi @idelgado Do you know about any updates for this feature?

**Comment by acossta** [comment]
@davidvoiq Nico from Twilio here. I'm the Product Manager for the SDKs. Piping the call events through to the originating leg is high up on our radar but don't have a timeline to share yet. I'll let you know as we make progress on this.

**Comment by mbrada** [comment]
Still no progress in this issue? It's have been almost a year :-/

**Comment by vishu5367** [comment]
Hi, can you please provide a callback in your mobiles SDK's for being notified when an outgoing connection from the phone has been answered by the person being called? (something similar to the ConnectionListener).
Basically i Need a call duration between incoming and outgoing calls



Thanks

**Comment by kongminghan** [comment]
Hi, is there any update on this? I find difficulties in implementing Twilio mobile SDK without any connection state callback. Thanks

**Comment by michaelrkn** [comment]
@kbagchiGWC I'm having this same issue using the twilio.js client. I guess I need to establish a separate WebSocket connection to pass the statusCallbackEvent to the client, but this adds a lot of complexity to the architecture, when the call status should be able to be passed back easily through the WebSocket connection already established by twilio.js. 

I thought that the new customParameters feature would help with this, but for outbound calls, it only supports parameters being sent from the client, to the TwiML app. If I could send custom parameters (containing the outbound call status) back from the TwiML app to the client, this would be pretty simple. 

It's frustrating that there isn't a straightforward way to get information back about outbound calls with the Twilio SDKs currently. Thanks, @kbagchiGWC.

**Comment by kongminghan** [comment]
This issue has been around for more than 2 years? Did your team plan to implement this issue? If not my team would like to consider move out from Twilio.

**Comment by idelgado** [comment]
@michaelrkn I believe what you are describing was added in 1.4.25. Here is the changelog describing how to enable the feature. It requires setting a flag and making a change to TwiML. https://www.twilio.com/docs/voice/client/javascript/changelog#feature-support

@kongminghan we are addressing this on Android and iOS as well. The 3.X preview releases introduce "ringing", which when coupled with `answerOnBridge=true` on the TwiML side also provides the "answered" behavior you are looking for. You can find more information about this here https://www.twilio.com/docs/voice/voip-sdk/android/3x-changelog#api-changes_1 as well and checkout our quickstart for 3.x here: https://github.com/twilio/voice-quickstart-android/tree/3.x. I encourage you to try the 3.X preview. We are moving to beta very soon and will GA in Q4 of this year.

**Comment by michaelrkn** [comment]
@idelgado Oh, that makes complete sense. That will let me differentiate between answered and unanswered calls. But what about differentiating between unanswered, busy, and failed calls - is there any way to do that through twilio.js? (Sorry to hijack this thread; feel free to move this elsewhere.)

**Comment by ryan-rowland** [comment]
@michaelrkn Currently any unanswered, busy or failed calls will disconnect in the same way. If the recipient is a PSTN, it may play early media that indicates the call failed or the line was busy, but there's no indicator in the SDK itself.

**Comment by michaelrkn** [comment]
@ryan-rowland Got it. This is where being able to pass customer parameters back *from* the TwiML *to* the twilio.js that initiated the outbound call would be really helpful - the final call status could be passed back this way. Thanks!
